README.md
# Arangodantic

[![GitHub Workflow Status](https://img.shields.io/github/workflow/status/ioxiocom/arangodantic/Build%20and%20upload%20to%20PyPI)](https://github.com/ioxiocom/arangodantic/actions/workflows/publish.yaml)
[![Code style: black](https://img.shields.io/badge/code%20style-black-000000.svg)](https://github.com/psf/black)
[![PyPI](https://img.shields.io/pypi/v/arangodantic)](https://pypi.org/project/arangodantic/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/arangodantic)](https://pypi.org/project/arangodantic/)
[![License: BSD 3-Clause](https://img.shields.io/badge/License-BSD%203--Clause-blue.svg)](https://opensource.org/licenses/BSD-3-Clause)

Database models for ArangoDB using Pydantic base models.

## Installation

The package is available on PyPI:

```bash
pip install arangodantic

# Or with Shylock
pip install arangodantic[shylock]
```

## Usage

Define your database models by extending either `DocumentModel` or `EdgeModel`. Nested
structures can be created by extending pydantic.BaseModel.

Configure Arangodantic. You can optionally define a collection name prefix, a key
generation function and a lock (needed if you want to use the locking functionality;
[Shylock](https://github.com/lietu/shylock) is supported out of the box, any other
locking service such as [Sherlock](https://pypi.org/project/sherlock/) should at least
in theory also work).

Ensure you have an ArangoDB server available with known credentials

```bash
docker run --rm -p 8529:8529 -e ARANGO_ROOT_PASSWORD="" arangodb/arangodb:3.7.2.1
```

```python
import asyncio
from uuid import uuid4

from aioarangodb import ArangoClient
from pydantic import BaseModel
from shylock import AsyncLock as Lock
from shylock import ShylockAioArangoDBBackend
from shylock import configure as configure_shylock

from arangodantic import ASCENDING, DocumentModel, EdgeModel, configure
from arangodantic.indexes import HashIndex


# Define models
class Owner(BaseModel):
    """Dummy owner Pydantic model."""

    first_name: str
    last_name: str


class Company(DocumentModel):
    """Dummy company Arangodantic model."""

    class ArangodanticConfig:
        indexes = [
            HashIndex(fields=["company_id"], unique=True),
            HashIndex(fields=["owner.first_name"]),
            HashIndex(fields=["owner.last_name"]),
        ]

    company_id: str
    owner: Owner


class Link(EdgeModel):
    """Dummy Link Arangodantic model."""

    class ArangodanticConfig:
        indexes = [HashIndex(fields=["_from", "_to", "type"], unique=True)]

    type: str


async def main():
    # Configure the database settings
    hosts = "http://localhost:8529"
    username = "root"
    password = ""
    database = "example"
    prefix = "example-"

    client = ArangoClient(hosts=hosts)
    # Connect to "_system" database and create the actual database if it doesn't exist
    # Only for demo, you likely want to create the database in advance.
    sys_db = await client.db("_system", username=username, password=password)
    if not await sys_db.has_database(database):
        await sys_db.create_database(database)

    # Configure Arangodantic and Shylock
    db = await client.db(database, username=username, password=password)
    configure_shylock(await ShylockAioArangoDBBackend.create(db, f"{prefix}shylock"))
    configure(db, prefix=prefix, key_gen=uuid4, lock=Lock)

    # Create collections and indexes if they don't yet exist
    # Only for demo, you likely want to create the collections in advance.
    await Company.ensure_collection()
    await Company.ensure_indexes()
    await Link.ensure_collection()
    await Link.ensure_indexes()

    # Clean up any existing data.
    await Company.truncate_collection()
    await Link.truncate_collection()

    # Let's create some example entries
    owner = Owner(first_name="John", last_name="Doe")
    company = Company(company_id="1234567-8", owner=owner)
    await company.save()
    print(f"Company saved with key: {company.key_}")

    second_owner = Owner(first_name="Jane", last_name="Doe")
    second_company = Company(company_id="2345678-9", owner=second_owner)
    await second_company.save()
    print(f"Second company saved with key: {second_company.key_}")

    link = Link(_from=company, _to=second_company, type="CustomerOf")
    await link.save()
    print(f"Link saved with key: {link.key_}")

    # Hold named locks while loading and doing changes
    async with Company.lock_and_load(company.key_) as c:
        assert c.owner == owner
        c.owner.first_name = "Joanne"
        await c.save()

    await company.reload()
    print(f"Updated owner of company to '{company.owner!r}'")

    # Let's explore the find functionality
    # Note: You likely want to add indexes to support the queries
    print("Finding companies owned by a person with last name 'Doe'")
    async with (await Company.find({"owner.last_name": "Doe"}, count=True)) as cursor:
        print(f"Found {len(cursor)} companies")
        async for found_company in cursor:
            print(f"Company: {found_company.company_id}")

    # Supported operators include: "==", "!=", "<", "<=", ">", ">="
    found_company = await Company.find_one(
        {"owner.last_name": "Doe", "_id": {"!=": company}}
    )
    print(f"Found the company {found_company.key_}")

    # Find also supports sorting and the cursor can easily be converted to a list
    companies = await (
        await Company.find(
            sort=[
                ("owner.last_name", ASCENDING),
                ("owner.first_name", ASCENDING),
            ]
        )
    ).to_list()
    print("Companies sorted by owner:")
    for c in companies:
        print(f"Company {c.company_id}, owner: {c.owner!r}")


if __name__ == "__main__":
    # Starting from Python 3.7 ->
    # asyncio.run(main())

    # Compatible with Python 3.6 ->
    loop = asyncio.get_event_loop()
    result = loop.run_until_complete(main())
```

You might find [migrate-anything](https://github.com/cocreators-ee/migrate-anything)
useful for creating and managing collections and indexes.

## More examples

- The [graph example](examples/graph_example.py) shows how arangodantic can be used with
  graphs. Please note that graph related functionality is at the moment really limited
  and will likely be extended later and might even be restructured, so use with caution.

## License

This code is released under the BSD 3-Clause license. Details in the
[LICENSE](./LICENSE) file.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.1.0/), and
this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [0.4.0] - 2023-09-12
- Switch to [Pydantic v2](https://docs.pydantic.dev/latest/)
- switch to python-arango because aioarango not maintained since 2 years
- add asyncer for [async](https://asyncer.tiangolo.com/) code calls
- rewrite tests to asyncer

### Todo 0.4.0
- add poetry
- rework tasks.py (neccessary with ruff?)
- rework setup.cfg (neccessary with ruff?)

## [0.3.1] - 2022-08-05

### Changed

- Switch from Travis CI to GitHub actions.

### Added

- Add back support for Python 3.7.

## [0.3.0] - 2022-08-04

### Removed

- Remove support for Python 3.7.

### Changed

- Update dependencies.
- Updates due to rename of GitHub organization.

## [0.2.0] - 2021-06-22

### Added

- Add `sort` parameter to `find` and `find_one` methods in models.
- Add `full_count` property to `ArangodanticCursor`.
- Add `offset` parameter to be used together with `limit` in `find`.
- Add `from_key_` and `to_key_` properties to edges.

### Changed

- Saving (especially edge models) through the graph can now raise a `ModelNotFoundError`
  instead of `DocumentInsertError` in certain cases.
- Update `pydantic` to ^1.8.2 that fixes CVE-2021-29510.

## [0.1.0] - 2020-10-27

### Added

- Add "find" and "find_one" methods to models.
- Add "truncate_collection" method to models.

### Changed

- The "delete_collection" method now raises DataSourceNotFound if the collection is
  missing and ignore_missing is set to False.
- Bump version to 0.1.0, project seems to be stable enough and has enough features for
  that.

## [0.0.4] - 2020-10-21

### Added

- Initial support for graphs

### Changed

- Renamed _\_before_save_ to _before_save_
- Bump version

## [0.0.3] - 2020-10-14

### Changed

- Update README.md
- Support for unique constraint errors
- Refactor handling of locks
- Bump version

## [0.0.2] - 2020-10-09

### Changed

- Update deployment credentials
- Bump version

## [0.0.1] - 2020-10-09

### Added

- Initial release of Arangodantic

[unreleased]: https://github.com/ioxiocom/arangodantic/compare/0.3.1...HEAD
[0.3.1]: https://github.com/ioxiocom/arangodantic/compare/0.3.0...0.3.1
[0.3.0]: https://github.com/ioxiocom/arangodantic/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/ioxiocom/arangodantic/compare/0.1.0...0.2.0
[0.1.0]: https://github.com/ioxiocom/arangodantic/compare/0.0.4...0.1.0
[0.0.4]: https://github.com/ioxiocom/arangodantic/compare/0.0.3...0.0.4
[0.0.3]: https://github.com/ioxiocom/arangodantic/compare/0.0.2...0.0.3
[0.0.2]: https://github.com/ioxiocom/arangodantic/compare/0.0.1...0.0.2
[0.0.1]: https://github.com/ioxiocom/arangodantic/releases/tag/0.0.1


examples/readme_example.py
import asyncio
from os import getenv
from uuid import uuid4

import asyncer
from arango import ArangoClient
from asyncer import asyncify
from orjson import orjson
from pydantic import BaseModel
from shylock import AsyncLock as Lock
from shylock import configure as configure_shylock

from arangodantic import ASCENDING, DocumentModel, EdgeModel, configure
from arangodantic.backends.asyncer_python_arango_backend import ShylockAsyncerArangoDBBackend


# Define models
class Owner(BaseModel):
    """Dummy owner Pydantic model."""

    first_name: str
    last_name: str


class Company(DocumentModel):
    """Dummy company Arangodantic model."""

    company_id: str
    owner: Owner


class Link(EdgeModel):
    """Dummy Link Arangodantic model."""

    key_: str = None
    type: str

    # @field_serializer("key_")
    # def serialize_key(self, key, _info):
    #     if key is None:
    #         return
    #     elif isinstance(key, uuid.UUID):
    #         return str(uuid)
    #     else:
    #         return key


async def main():
    # Configure the database settings
    hosts = getenv("HOSTS")
    username = getenv("USERNAME")
    password = getenv("PASSWORD")
    database = "example"
    prefix = "example-"
    # Create the ArangoDB client with orjson as serializer and deserializer
    client = ArangoClient(
        hosts=hosts, serializer=orjson.dumps, deserializer=orjson.loads
    )
    # Connect to "_system" database and create the actual database if it doesn't exist
    # Only for demo, you likely want to create the database in advance.
    sys_db = await asyncify(client.db)("_system", username=username, password=password)
    if not await asyncify(sys_db.has_database)(database):
        await asyncify(sys_db.create_database)(database)

    # Configure Arangodantic and Shylock
    db = await asyncify(client.db)(database, username=username, password=password)
    configure_shylock(
        await ShylockAsyncerArangoDBBackend.create(db, f"{prefix}shylock")
    )
    configure(db, prefix=prefix, key_gen=uuid4, lock=Lock)

    # Create collections if they don't yet exist
    # Only for demo, you likely want to create the collections in advance.
    async with asyncer.create_task_group() as tg:
        tg.soonify(Company.ensure_collection)()
        tg.soonify(Link.ensure_collection)()

    # Let's create some example entries
    owner = Owner(first_name="John", last_name="Doe")
    company = Company(company_id="1234567-8", owner=owner)
    await company.save()
    print(f"Company saved with key: {company.key_}")

    second_owner = Owner(first_name="Jane", last_name="Doe")
    second_company = Company(company_id="2345678-9", owner=second_owner)
    await second_company.save()
    print(f"Second company saved with key: {second_company.key_}")

    link = Link(_from=company, _to=second_company, type="CustomerOf")
    print(link)
    await link.save()
    print(f"Link saved with key: {link.key_}")

    # Hold named locks while loading and doing changes
    async with Company.lock_and_load(company.key_) as c:
        assert c.owner == owner
        c.owner.first_name = "Joanne"
        await c.save()

    await company.reload()
    print(f"Updated owner of company to '{company.owner!r}'")

    # Let's explore the find functionality
    # Note: You likely want to add indexes to support the queries
    print("Finding companies owned by a person with last name 'Doe'")
    # async with asyncer.create_task_group() as tg:
    #     soon = tg.soonify(Company.find)({"owner.last_name": "Doe"}, count=True)

    companies = await (
        await Company.find({"owner.last_name": "Doe"}, count=True)
    ).to_list()
    print(f"Found {len(companies)} companies")
    # Cave return as dict at the moment
    # Todo return as list of models
    print(f"Found companies\n {companies}")

    # Supported operators include: "==", "!=", "<", "<=", ">", ">="
    found_company = await Company.find_one(
        {"owner.last_name": "Doe", "_id": {"!=": company}}
    )
    print(f"Found the company {found_company.key_}")
    print(f"Found the company {type(found_company)}")

    # Find also supports sorting and the cursor can easily be converted to a list

    companies = await (
        await Company.find(
            sort=[("owner.last_name", ASCENDING), ("owner.first_name", ASCENDING)]
        )
    ).to_list()
    print(f"comps {companies}")
    print(f"type {type(companies)}")
    print("Companies sorted by owner:")
    for c in companies:
        company = Company(**c)
        print(f"Company {company.company_id}, owner: {company.owner!r}")


if __name__ == "__main__":
    asyncio.run(main())
